modules/telemetry/json_packets.py

```diff
@@ -185,3 +191,170 @@ def parse_mission_file(mission_file: Path) -> MissionEntry:
             length += 1
 
     return MissionEntry(name=mission_file.stem, length=length, filepath=mission_file, version=1)
+
+
+@dataclass
+class TelemetryDataPacketBlock:
+    """A generic block object to store information for telemetry data
+    All stored values must be updated at once!"""
+
+    mission_time: list[int] = field(default_factory=list[int])
+    stored_values: dict[str, list[int]] = field(default_factory=dict)
+
+    def update(self, data: dict[str, int], buffer_size: int) -> None:
+        """Updates the stored values with the given data
+        Args:
+            data (dict[str, int]) : Dictionary of data to update containing mission_time and stored_values squashed
+            buffer_size (int) : Size of the telemetry buffer"""
+
+        # Ensure you are not half updating the packet
+        # As this can cause the arrays to become out of sync and meaningless.
+        if dict(self).keys() != data.keys():
+            logger.error("Block must be updated using a full set of values at the same time!")
+            logger.debug(f"Tried updating {dict(self).keys()} using {data.keys()}!")
+            return
+
+        # Updates stored values with new values
+        for key in data.keys():
+            if key == "mission_time":
+                self.mission_time.append(data["mission_time"])
+            else:
+                self.stored_values[key].append(data[key])
+
+        # Ensure buffer sizes are kept
+        # Note: This uses a while loop incase buffer size shrinks drastically
+        while len(self.mission_time) > buffer_size:
+            self.mission_time.pop(0)
+            for key in self.stored_values.keys():
+                self.stored_values[key].pop(0)
+
+    def clear(self) -> None:
+        """Clears all stored values"""
+        self.mission_time = []
+        self.stored_values = {key: [] for key in self.stored_values.keys()}
+
+    def __str__(self):
+        """Returns a string representation of the TelemetryDataPacketBlock"""
+        return f"{self.__class__.__name__} -> time: {self.mission_time} ms, {self.stored_values}"
+
+    def __iter__(self):
+        """Returns an interator containing all the stored values"""
+        yield "mission_time", self.mission_time
+        for key in self.stored_values.keys():
+            yield key, self.stored_values[key]
+
+
+class TelemetryData:
+    """Contains the output specification for the telemetry data block"""
+
+    # Configuration
+    buffer_size: int
+    decoder: list[dict[int, dict[str, str]]]
+
+    # Current data storage
+    last_mission_time: int
+    output_blocks: dict[str, TelemetryDataPacketBlock]
+    update_buffer: dict[str, dict[str, float | int | str | None]]
```

Comment:
None of these serve a purpose, they're stub declarations for class attributes. I mean you need to delete all of these and only define them in `__init__` as instance attributes.

Reply:
Resolved in 5180605.